### PR TITLE
chore(flow): craft explicit flow types for notebook

### PR DESCRIPTION
**Pull request by rgbkrk**
I'm going to start getting our state to be a bit more exact to make typing of our reducers, actions, and epics much easier. This starts out by creating flow types for the overall notebook state.

### CODE REVIEW

**Comment by rgbkrk** [comment]
This last commit is going to fail due to `flow`. 😢  Immutable's `Record` is missing a definition for the constructor call of the "class"es it returns.

**Comment by rgbkrk** [comment]
Now for a commit that fails in the other direction:

```
src/notebook/index.js:33
 33:   document: DocumentRecord(),
                 ^^^^^^^^^^^^^^^^ function call. Callable signature not found in
 33:   document: DocumentRecord(),
                 ^^^^^^^^^^^^^^ Record
```

**Comment by rgbkrk** [comment]
Since `Record`'s returned objects also don't have a Callable signature, also stuck there. 😢

**Comment by rgbkrk** [comment]
and this last commit will pass.

**Comment by rgbkrk** [comment]
If I set the container type for a Document Record as:

```
export const DocumentRecord: Immutable.Record<Document> = Immutable.Record({
  notebook: null,
  transient: new Immutable.Map({
    keyPathsForDisplays: new Immutable.Map(),
  }),
  cellPagers: new Immutable.Map(),
  outputStatuses: new Immutable.Map(),
  stickyCells: new Immutable.Set(),
  editorFocused: null,
  cellFocused: null,
  copied: new Immutable.Map(),
});
```

that's when it's unable to determine that it's a Callable. Any ideas for workarounds @peggyrayzis?

**Comment by rgbkrk** [comment]
Solid SO answer here about why: http://stackoverflow.com/questions/37033339/how-to-describe-immutable-js-map-shape-with-flow

**Comment by rgbkrk** [comment]
This makes me want to switch off of Immutable because I love `flow` so much, yet there are no other good libraries in this space that have a good `flow` story either (playing with `lodash/fp` taught me this).

**Comment by rgbkrk** [comment]
However, with `lodash/fp` these type definitions would match really well.

**Comment by rgbkrk** [comment]
Interesting way to do it that lets you use Records in the way we expect:

```
export type DocumentSpec = {
  notebook: Notebook,
  transient: Immutable.Map<string, any>,
  cellPagers: any,
  outputStatuses: Immutable.Map<string, any>,
  stickyCells: Immutable.Set<any>,
  editorFocused: any,
  cellFocused: any,
  copied: Immutable.Map<any, any>,
}

export type Document = DocumentSpec & Immutable.Record<DocumentSpec>

export const DocumentRecord: Document = Immutable.Record({
  notebook: null,
  transient: new Immutable.Map({
    keyPathsForDisplays: new Immutable.Map(),
  }),
  cellPagers: new Immutable.Map(),
  outputStatuses: new Immutable.Map(),
  stickyCells: new Immutable.Set(),
  editorFocused: null,
  cellFocused: null,
  copied: new Immutable.Map(),
});
```

Still can't have the Callable signature.

Also `Function cannot be called on any member of intersection type`

File: src/notebook/records.js [hunk 1]
```diff
@@ -14,7 +117,18 @@ export const AppRecord = new Immutable.Record({
...
 });
 
-export const DocumentRecord = new Immutable.Record({
+export type Document = {
+  notebook: Notebook,
+  transient: Immutable.Map<string, any>,
+  cellPagers: any,
+  outputStatuses: Immutable.Map<string, any>,
+  stickyCells: Immutable.Set<any>,
+  editorFocused: any,
```
Comment (peggyrayzis) [inline src/notebook/records.js:126]: I think you can mark editorFocused & cellFocused as string|null here.

**Review by peggyrayzis — APPROVED** [review]
What you have now serves our purpose well, although it sucks that we can't define the object as a Callable. Honestly, I need to look into the typing of Callable objects closer to suggest any other possible workarounds. Do you think it would be worth working on a PR to the flow-typed immutable libdef to make records callable?

**Comment by peggyrayzis** [comment]
Also, thank you for typing all of this. I like having the types in records, I think it makes the code more discoverable 😀

**Comment by rgbkrk** [comment]
> Also, thank you for typing all of this. I like having the types in records, I think it makes the code more discoverable 😀

You make me feel more positive about including this since at the very least the types can be used with the reducers.

**Comment by rgbkrk** [comment]
> Do you think it would be worth working on a PR to the flow-typed immutable libdef to make records callable?

The libdef for Immutable is in the Immutable.js repo. The number of open PRs attempting to fix the flow definitions as well as the unshipped commits make me worried about whether there's a good way to even go about this.

**Comment by rgbkrk** [comment]
Alright, pushed a commit that at least passes and will be useful for people to see/use.

**Comment by peggyrayzis** [comment]
Yeah, I think it will definitely be useful as we start to type the reducers. Is this still WIP?

**Comment by rgbkrk** [comment]
Good to go!

**Comment by rgbkrk** [comment]
I'm going to be trying out https://www.npmjs.com/package/flow-immutable-models with jscodeshift today, seeing how the code generation works out.

**Comment by peggyrayzis** [comment]
Keep me updated, it looks awesome!

**Pull request merged**